homeassistant/helpers/intent.py

```diff
@@ -117,6 +124,72 @@ def __repr__(self):
         return '<{} - {}>'.format(self.__class__.__name__, self.intent_type)
 
 
+def fuzzymatch(name, entities):
+    """Fuzzy matching function."""
+    matches = []
+    pattern = '.*?'.join(name)
+    regex = re.compile(pattern)
+    for entity in entities:
+        match = regex.search(entity)
+        if match:
+            matches.append((len(match.group()), match.start(), entity))
+    return [x for _, _, x in sorted(matches)]
+
+
+class ServiceIntentHandler(IntentHandler):
+    """Service Intent handler registration.
+
+    Service specific intent handler that calls a service by name/entity_id.
+    """
+
+    slot_schema = {
+        'name': cv.string,
+    }
+
+    def __init__(self, intent_type, domain, service, speech):
+        """Create Service Intent Handler."""
+        self.intent_type = intent_type
+        self.domain = domain
+        self.service = service
+        self.speech = speech
+
+    @asyncio.coroutine
+    def async_handle(self, intent_obj):
+        """Handle the hass intent."""
+        hass = intent_obj.hass
+        slots = self.async_validate_slots(intent_obj.slots)
+        response = intent_obj.create_response()
+
+        name = slots['name']['value']
+        entities = {state.entity_id: state.name for state
+                    in hass.states.async_all()}
+        entity_name = name.replace(' ', '_').lower()
+        entity_name = entity_name.replace('the_', '')
+
+        matches = fuzzymatch(entity_name, entities)
+        entity_id = next((x for x in matches if self.domain in x), None)
+        if not entity_id:
+            entity_id = matches[0] if matches else None
+        _LOGGER.debug("%s matched entity: %s", name, entity_id)
+
+        response = intent_obj.create_response()
+        if not entity_id:
+            response.async_set_speech(
+                "Could not find entity id matching {}".format(name))
+            _LOGGER.error("Could not find entity id matching %s (%s)", name,
+                          entity_name)
+            return response
+
+        yield from hass.services.async_call(
+            self.domain, self.service, {
+                ATTR_ENTITY_ID: entity_id
+            }, blocking=True)
```

Comment:
Let's remove blocking, it has been causing trouble for Alexa/Google Assistant as turning on can sometimes take a long time and we don't actually know if a service is successful.